### ISSUE TITLE
Intersection of ImageSet gives incorrect answer.

### ISSUE THREAD

**Issue by jmig5776** [sympy/sympy#17355]
After git bisecting by @gschintgen this [commit ](https://github.com/sympy/sympy/commit/f54aa8d4593bbc107af91f6f033a363dd3a440db) has changed the output of 
```python
>>> Intersection(S.Integers, ImageSet(Lambda(n, 5*n + 3), S.Integers))
S.Integers
# expected ImageSet(Lambda(n, 5*n + 3), S.Integers)
```
ping - @smichr

**Comment by asmeurer** [comment]
That commit is from this PR https://github.com/sympy/sympy/pull/16864

**Comment by oscarbenjamin** [comment]
This is a regression since 1.4

**Comment by oscarbenjamin** [comment]
It goes wrong here:
https://github.com/sympy/sympy/blob/21183076095704d7844a832d2e7f387555934f0c/sympy/sets/handlers/intersection.py#L231
I'm not sure what that code is trying to do but we only hit that branch when calculating the intersection of the an ImageSet with Integers as base set and the Integers so at that point the result of the intersection is just self.

**Comment by oscarbenjamin** [comment]
There is a fix in #17813
